pkg/apis/pipeline/v1beta1/param_types.go

```diff
@@ -107,17 +132,23 @@ type ArrayOrString struct {
 	Type      ParamType `json:"type"` // Represents the stored type of ArrayOrString.
 	StringVal string    `json:"stringVal"`
 	// +listType=atomic
-	ArrayVal []string `json:"arrayVal"`
+	ArrayVal  []string          `json:"arrayVal"`
+	ObjectVal map[string]string `json:"objectVal"`
 }
 
 // UnmarshalJSON implements the json.Unmarshaller interface.
 func (arrayOrString *ArrayOrString) UnmarshalJSON(value []byte) error {
-	if value[0] == '"' {
+	switch value[0] {
```

Comment:
I realized we probably need to check the value is not empty? Otherwise get index out of bound error

Reply:
Seems like this [original code](https://github.com/tektoncd/pipeline/blob/a3caabf08751b3a8515bf2e8528ee728c49280ae/pkg/apis/pipeline/v1beta1/param_types.go#L113-L121) didn't check that. And I guess the reason is that no matter what users provided in yaml, k8s will convert it to json so `value` will not be empty even though users provided nothing for a field i.e. `default: `

Otherwise, can you show me an example yaml that might trigger index out of bound error?

Reply:
I will add this in results's PR

Reply:
sg! if this is specific to Unmarshal issue on the result side, it seems to make more sense to add that check behaviour in relevant pr so that people can track why.

Reply:
https://github.com/tektoncd/pipeline/pull/4839 might be related!